Add a feature to provide each process the generated ports from other processes

offbyone opened the pull request:
- include the ports generated for other processes in each process definition
- Add `OVERMIND_PROCESS_<name>_PORT` settings to each process wrapper script
- add docs for the feature

Implements #189

offbyone committed:
Add a feature to provide each process the generated ports from other processes

- include the ports generated for other processes in each process
  definition
- Add `OVERMIND_PROCESS_<name>_PORT` settings to each process wrapper
  script
- add docs for the feature

DarthSim commented:
Hey @offbyone!

Thanks for your contribution!

Here are a few things I would change before merging this:

1. I don't think we need a flag that disables this behavior. I don't see a case where this would be required, and I don't want to occupy a short flag name.

2. We don't need `procfileEntry.OtherProcessPorts` field since we already have entries list. We can pass it to `createScriptFile`.

3. Process names may contain symbols not allowed in environment variable names. When building environment variable names, we should replace all non-alphanumeric symbols with an underscore.

offbyone commented:
I'm happy to make all of those changes. I erred on the side of over-gating the feature but it should be easy to simplify.

offbyone commented:
Looking at my own shell and the Open Group's docs, it looks like `=` is the only character not permitted in an environment variable name:
 
```shellsession
$ env FAKE_😂=something env | grep FAKE
FAKE_😂=something
$ bash
bash-5.2$ env FAKE_😂=something env | grep FAKE
FAKE_😂=something
```

I was not able to make this fail on any machine I have access to.

offbyone committed:
Address PR feedback

- drop the extra command line flag and gate
- add a bit of extra safety to the port environment variable name

offbyone committed:
PR feedback: don't carry the other ports around in the procfile

Construct the other service port list as the script is generated

DarthSim commented:
> Looking at my own shell and the Open Group's docs, it looks like = is the only character not permitted in an environment variable name

The behavior differs from shell to shell. For example, Fish allows variable names to contain only alphanumeric symbols and underscores. I would rather not take risks here; it's better to be safe than sorry.

offbyone commented:
My example actually _was_ in fish shell. Here's further indications that fish handles environment (as opposed to shell) variables with emoji just fine:

```shellsession
$ set var_with_😂 "foo"
set: var_with_😂: invalid variable name. See `help identifiers`

(Type 'help set' for related documentation)
$ env ENV_VAR_WITH_😂=foo fish
Welcome to fish, the friendly interactive shell
Type help for instructions on how to use fish
$ env | grep ENV_VAR_
ENV_VAR_WITH_😂=foo
```

offbyone commented:
(that said, if you want me to, I still can carve those out, I don't mind, but I don't think it's an issue here)

offbyone commented:
Sounds good. I'll strip them down to `[A-Za-z0-9_]`

offbyone committed:
Explicitly allow only [a-zA-Z0-9_] in the process names

DarthSim commented on start/command.go:
```diff
@@ -247,3 +256,7 @@ func (c *command) waitForTimeoutOrStop() {
 }
+
+func sanitizeProcName(name string) string {
+	return regexp.MustCompile(`[^a-zA-Z0-9]`).ReplaceAllString(name, "_")
```
A couple of tiny things here:

1. For readability sake, I'd add `_` to the regexp
2. I'd put the regexp into a global var to not compile it on every func call

offbyone committed:
Precompile the process names

DarthSim commented:
Thanks again for your contribution!

dobesv commented:
Was this feature released?  It doesn't seem to be working for me, and the release notes for releases around this time do not seem to mention it.

rolodato commented:
It was not released. The code from the latest release does not have this feature: https://github.com/DarthSim/overmind/blob/v2.5.1/start/command.go#L88

Example:

```
➜ overmind --version
Overmind version 2.5.1
➜ cat Procfile
p1: echo port is $PORT; sleep 5
p2: echo p1 port is $OVERMIND_PROCESS_p1_PORT
➜ overmind start
system | Tmux socket name: overmind-sandbox-Srjr-mHisG-iQUFrpDh1C
system | Tmux session ID: sandbox
system | Listening at ./.overmind.sock
p2     | Started with pid 44907...
p1     | Started with pid 44906...
p2     | p1 port is
p2     | Exited with code 0
p1     | Interrupting...
p1     | Exited with code 0
```

@DarthSim This is a really useful feature and would love to get it released. Thank you!

offbyone commented:
I'm still hoping this will be released as well. In the meantime, I've got a build on my fork with this change in it; I'll be cherry picking other PRs as they come in: https://github.com/offbyone/overmind/releases/tag/v2.5.2-fork.1
